# Overview

This notebook presents step by step explanation of a pipeline designed to collect event updates from an API, clean, transform and load them to the database. The data is collected from the GDELT repository. It includes the world's broadcast, print, and web news from nearly every corner of every country in over 100 languages and identifies the people, locations, organizations, themes, sources, emotions, counts, quotes, images and events driving our global society every second of every day, creating a free open platform for computing on the entire world.

In [91]:
import pandas as pd

# Download zip file

In [92]:
import os
from urllib.request import urlopen
from urllib.request import urlretrieve
from zipfile import ZipFile

In [104]:
data = urlopen('http://data.gdeltproject.org/gdeltv2/lastupdate.txt')
data

In [94]:
for line in data:
    target_file_link = line
    break
    
target_file_link

b'133110 8051e8a2b93d6a80c6b410a03c5a9e4d http://data.gdeltproject.org/gdeltv2/20200617010000.export.CSV.zip\n'

In [95]:
target_link = target_file_link.decode("utf-8").split(" ")[2]
target_file_url = target_link
target_file_url

'http://data.gdeltproject.org/gdeltv2/20200617010000.export.CSV.zip\n'

In [96]:
target_file = target_link.split("/")[-1]
target_filename = target_file.replace(".zip\n", "")

In [97]:
target_file = target_filename

In [98]:
urlretrieve(target_file_url, "C:\\Users\\arali\\Files\\Professional\\Self Study\\Data Sceince\\Portfolio\\GDELT Pipeline - Desktop\\jupyter\\" + target_file + ".zip")

('C:\\Users\\arali\\Files\\Professional\\Self Study\\Data Sceince\\Portfolio\\GDELT Pipeline - Desktop\\jupyter\\20200617010000.export.CSV.zip',
 <http.client.HTTPMessage at 0x1a59c080c08>)

# Unzip download file

In [99]:
filename = 'C:\\Users\\arali\\Files\\Professional\\Self Study\\Data Sceince\\Portfolio\\GDELT Pipeline - Desktop\\jupyter\\' + target_file + '.zip'

with ZipFile(filename, 'r') as zip:
    # extracting all the files
    print('Extracting all the files now...')
    zip.extractall('C:\\Users\\arali\\Files\\Professional\\Self Study\\Data Sceince\\Portfolio\\GDELT Pipeline - Desktop\\jupyter\\')
    print('Done!')

Extracting all the files now...
Done!


# CSV to DataFrame

In [100]:
dataframe = pd.read_csv(
    'C:\\Users\\arali\\Files\\Professional\\Self Study\\Data Sceince\\Portfolio\\GDELT Pipeline - Desktop\\jupyter\\' +
    target_file,
    sep='\t',
    header=None,
    encoding='utf-8')


primary_fields = [
        'GLOBALEVENTID',
        'SQLDATE',
        'MonthYear',
        'Year',
        'FractionDate',
        'Actor1Code',
        'Actor1Name',
        'Actor1CountryCode',
        'Actor1KnownGroupCode',
        'Actor1EthnicCode',
        'Actor1Religion1Code',
        'Actor1Religion2Code',
        'Actor1Type1Code',
        'Actor1Type2Code',
        'Actor1Type3Code',
        'Actor2Code',
        'Actor2Name',
        'Actor2CountryCode',
        'Actor2KnownGroupCode',
        'Actor2EthnicCode',
        'Actor2Religion1Code',
        'Actor2Religion2Code',
        'Actor2Type1Code',
        'Actor2Type2Code',
        'Actor2Type3Code',
        'IsRootEvent',
        'EventCode',
        'EventBaseCode',
        'EventRootCode',
        'QuadClass',
        'GoldsteinScale',
        'NumMentions',
        'NumSources',
        'NumArticles',
        'AvgTone',
        'Actor1Geo_Type',
        'Actor1Geo_FullName',
        'Actor1Geo_CountryCode',
        'Actor1Geo_ADM1Code',
        'gap_1',
        'Actor1Geo_Lat',
        'Actor1Geo_Long',
        'Actor1Geo_FeatureID',
        'Actor2Geo_Type',
        'Actor2Geo_FullName',
        'Actor2Geo_CountryCode',
        'gap_2',
        'Actor2Geo_ADM1Code',
        'Actor2Geo_Lat',
        'Actor2Geo_Long',
        'Actor2Geo_FeatureID',
        'ActionGeo_Type',
        'ActionGeo_FullName',
        'ActionGeo_CountryCode',
        'ActionGeo_ADM1Code',
        'gap_3',
        'ActionGeo_Lat',
        'ActionGeo_Long',
        'ActionGeo_FeatureID',
        'DATEADDED',
        'SOURCEURL']

dataframe.columns = primary_fields
dataframe

GLOBALEVENTID   SQLDATE  MonthYear  Year  FractionDate Actor1Code  \
0         929813247  20190618     201906  2019     2019.4603        CVL   
1         929813248  20190618     201906  2019     2019.4603        GBR   
2         929813249  20190618     201906  2019     2019.4603        GBR   
3         929813250  20190618     201906  2019     2019.4603        GBR   
4         929813251  20190618     201906  2019     2019.4603        GBR   
...             ...       ...        ...   ...           ...        ...   
2015      929815262  20200617     202006  2020     2020.4575        iku   
2016      929815263  20200617     202006  2020     2020.4575        ltn   
2017      929815264  20200617     202006  2020     2020.4575        ltn   
2018      929815265  20200617     202006  2020     2020.4575        ltn   
2019      929815266  20200617     202006  2020     2020.4575        may   

          Actor1Name Actor1CountryCode Actor1KnownGroupCode Actor1EthnicCode  \
0          COMMUNITY               NaN                  NaN              NaN   
1     UNITED KINGDOM               GBR                  NaN              NaN   
2     UNITED KINGDOM               GBR                  NaN              NaN   
3     UNITED KINGDOM               GBR                  NaN              NaN   
4     UNITED KINGDOM               GBR                  NaN              NaN   
...              ...               ...                  ...              ...   
2015           INUIT               NaN                  NaN              iku   
2016         LATINOS               NaN                  NaN              ltn   
2017         LATINOS               NaN                  NaN              ltn   
2018         LATINOS               NaN                  NaN              ltn   
2019          MALAYS               NaN                  NaN              may   

      ... ActionGeo_Type                      ActionGeo_FullName  \
0     ...              4  Morecambe Bay, Cumbria, United Kingdom   
1     ...              4   Carnforth, Lancashire, United Kingdom   
2     ...              4  Morecambe Bay, Cumbria, United Kingdom   
3     ...              5                     Anhui, Anhui, China   
4     ...              2               New Jersey, United States   
...   ...            ...                                     ...   
2015  ...              1                            South Africa   
2016  ...              2               California, United States   
2017  ...              2               California, United States   
2018  ...              2               California, United States   
2019  ...              1                                Malaysia   

     ActionGeo_CountryCode ActionGeo_ADM1Code  gap_3 ActionGeo_Lat  \
0                       UK               UKC9  40118       54.1167   
1                       UK               UKH2  40118       54.1167   
2                       UK               UKC9  40118       54.1167   
3                       CH               CH01  12996       32.0000   
4                       US               USNJ    NaN       40.3140   
...                    ...                ...    ...           ...   
2015                    SF                 SF    NaN      -30.0000   
2016                    US               USCA    NaN       36.1700   
2017                    US               USCA    NaN       36.1700   
2018                    US               USCA    NaN       36.1700   
2019                    MY                 MY    NaN        2.5000   

     ActionGeo_Long ActionGeo_FeatureID       DATEADDED  \
0          -3.00000            -2603426  20200617010000   
1          -2.76667            -2591885  20200617010000   
2          -3.00000            -2603426  20200617010000   
3         117.00000            -1896677  20200617010000   
4         -74.50890                  NJ  20200617010000   
...             ...                 ...             ...   
2015       26.00000                  SF  20200617010000   
2016     -119.

In [102]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2020 entries, 0 to 2019
Data columns (total 61 columns):
GLOBALEVENTID            2020 non-null int64
SQLDATE                  2020 non-null int64
MonthYear                2020 non-null int64
Year                     2020 non-null int64
FractionDate             2020 non-null float64
Actor1Code               1834 non-null object
Actor1Name               1834 non-null object
Actor1CountryCode        1087 non-null object
Actor1KnownGroupCode     18 non-null object
Actor1EthnicCode         9 non-null object
Actor1Religion1Code      15 non-null object
Actor1Religion2Code      10 non-null object
Actor1Type1Code          884 non-null object
Actor1Type2Code          72 non-null object
Actor1Type3Code          3 non-null object
Actor2Code               1427 non-null object
Actor2Name               1427 non-null object
Actor2CountryCode        820 non-null object
Actor2KnownGroupCode     21 non-null object
Actor2EthnicCode         6 non-null obje

Dataset contains 2020 events. Idenrifer and data columns are complete, but there are certain number of missing value in other columns. 

# Clean data

To clean the data number of missing values and relevance of the information is considered.

In [106]:
dataframe.Actor1Type1Code.unique()

array(['CVL', nan, 'COP', 'MIL', 'BUS', 'HLH', 'LAB', 'MED', 'GOV', 'EDU',
       'JUD', 'CRM', 'ELI', 'REF', 'IGO', 'OPP', 'LEG', 'MNC', 'NGO',
       'RAD', 'UAF'], dtype=object)

In [107]:
dataframe.Actor1Type1Code.nunique()

20

For this project, we consider only actor 1 as it has less missing values. Events with missing actor code are ignored and only relevant actor types are kept:

In [105]:
import pandasql as ps
sql_query = """
                    SELECT GLOBALEVENTID, CAST(SQLDATE AS INTEGER),
                    MonthYear, Year, Actor1Code, Actor1Type1Code,
                    ActionGeo_FullName, ActionGeo_ADM1Code,
                    Actor1Geo_CountryCode, CAST(GoldsteinScale AS FLOAT)
                    FROM dataframe
                    WHERE Actor1Geo_CountryCode='US' and
                    Actor1Code != 'null' and
                    Actor1Type1Code in ('COP', 'GOV', 'JUD', 'BUS',
                                        'CRM', 'DEV', 'EDU', 'ENV',
                                        'HLH', 'LEG','MED','MNC');
                    """

df = ps.sqldf(sql_query, locals())
df

GLOBALEVENTID  CAST(SQLDATE AS INTEGER)  MonthYear  Year Actor1Code  \
0        929813256                  20200518     202005  2020        COP   
1        929813257                  20200518     202005  2020        COP   
2        929813265                  20200610     202006  2020        BUS   
3        929813266                  20200610     202006  2020        BUS   
4        929813267                  20200610     202006  2020        BUS   
..             ...                       ...        ...   ...        ...   
348      929815208                  20200617     202006  2020     USAJUD   
349      929815209                  20200617     202006  2020     USAJUD   
350      929815210                  20200617     202006  2020     USALEG   
351      929815222                  20200617     202006  2020     USAMED   
352      929815223                  20200617     202006  2020     USAMED   

    Actor1Type1Code                          ActionGeo_FullName  \
0               COP                   New Jersey, United States   
1               COP                  Mississippi, United States   
2               BUS                 Tamaulipas, Sinaloa, Mexico   
3               BUS                        Texas, United States   
4               BUS                        Texas, United States   
..              ...                                         ...   
348             JUD                     Colorado, United States   
349             JUD                     Colorado, United States   
350             LEG                         Utah, United States   
351             MED                   California, United States   
352             MED  Monterey County, California, United States   

    ActionGeo_ADM1Code Actor1Geo_CountryCode  CAST(GoldsteinScale AS FLOAT)  
0                 USNJ                    US                            7.0  
1                 USMS                    US                           -6.5  
2                 MX25                    US                            3.0  
3                 USTX                    US                            3.0  
4                 USTX                    US                            5.0  
..                 ...                   ...                            ...  
348               USCO                    US                           -2.0  
349               USCO                    US                           -5.0  
350               USUT                    US                           -5.0  
351               USCA                    US                           -5.0  
352               USCA                    US                           -5.0  

[353 rows x 10 columns]

The number of events are reduced to 353. Furthermore, only events in specific states of the United States provides relevant  information. Unique geolocation code of the events existing in the dataset are obtained as follows:

In [59]:
df.ActionGeo_ADM1Code.unique()

array(['USCT', 'USCA', 'USTX', 'USFL', 'USUT', 'USDC', 'USMD', 'USIL',
       'USWI', 'USNY', 'US', 'USMO', 'USMN', 'USID', 'CO02', 'USAL',
       'USOR', 'USOK', 'USIA', 'USOH', 'USVA', 'USLA', 'USME', 'USGA',
       'IR', 'USIN', 'USMA', 'USNC', 'USKS', 'USTN', 'USWA', 'USHI',
       'USDE', 'USMI', 'USNH', 'USAZ', 'USKY', 'KN08', 'USNJ', 'USPA',
       'USAR', 'USNE', 'USMS', 'USMT', 'USCO', 'USNM', 'RS48', 'SP', 'RS',
       'KN12'], dtype=object)

The geolocation code has US in front of all states. Given that the purpose of the project is to investigate US states exclusively, state code is extracted by removing US from the front of the code.

In [110]:
action_state = df['ActionGeo_ADM1Code'].apply(lambda x: x[2:])
df['action_state'] = action_state
df

GLOBALEVENTID  CAST(SQLDATE AS INTEGER)  MonthYear  Year Actor1Code  \
0        929813256                  20200518     202005  2020        COP   
1        929813257                  20200518     202005  2020        COP   
2        929813265                  20200610     202006  2020        BUS   
3        929813266                  20200610     202006  2020        BUS   
4        929813267                  20200610     202006  2020        BUS   
..             ...                       ...        ...   ...        ...   
348      929815208                  20200617     202006  2020     USAJUD   
349      929815209                  20200617     202006  2020     USAJUD   
350      929815210                  20200617     202006  2020     USALEG   
351      929815222                  20200617     202006  2020     USAMED   
352      929815223                  20200617     202006  2020     USAMED   

    Actor1Type1Code                          ActionGeo_FullName  \
0               COP                   New Jersey, United States   
1               COP                  Mississippi, United States   
2               BUS                 Tamaulipas, Sinaloa, Mexico   
3               BUS                        Texas, United States   
4               BUS                        Texas, United States   
..              ...                                         ...   
348             JUD                     Colorado, United States   
349             JUD                     Colorado, United States   
350             LEG                         Utah, United States   
351             MED                   California, United States   
352             MED  Monterey County, California, United States   

    ActionGeo_ADM1Code Actor1Geo_CountryCode  CAST(GoldsteinScale AS FLOAT)  \
0                 USNJ                    US                            7.0   
1                 USMS                    US                           -6.5   
2                 MX25                    US                            3.0   
3                 USTX                    US                            3.0   
4                 USTX                    US                            5.0   
..                 ...                   ...                            ...   
348               USCO                    US                           -2.0   
349               USCO                    US                           -5.0   
350               USUT                    US                           -5.0   
351               USCA                    US                           -5.0   
352               USCA                    US                           -5.0   

    action_state  
0             NJ  
1             MS  
2             25  
3             TX  
4             TX  
..           ...  
348           CO  
349           CO  
350           UT  
351           CA  
352           CA  

[353 rows x 11 columns]

Furthermore, events related to general US are removed as they do not provide relevant info about specific state.

In [114]:
df1 = df.loc[df.ActionGeo_ADM1Code != "US"]

df1

GLOBALEVENTID  CAST(SQLDATE AS INTEGER)  MonthYear  Year Actor1Code  \
0        929813256                  20200518     202005  2020        COP   
1        929813257                  20200518     202005  2020        COP   
2        929813265                  20200610     202006  2020        BUS   
3        929813266                  20200610     202006  2020        BUS   
4        929813267                  20200610     202006  2020        BUS   
..             ...                       ...        ...   ...        ...   
348      929815208                  20200617     202006  2020     USAJUD   
349      929815209                  20200617     202006  2020     USAJUD   
350      929815210                  20200617     202006  2020     USALEG   
351      929815222                  20200617     202006  2020     USAMED   
352      929815223                  20200617     202006  2020     USAMED   

    Actor1Type1Code                          ActionGeo_FullName  \
0               COP                   New Jersey, United States   
1               COP                  Mississippi, United States   
2               BUS                 Tamaulipas, Sinaloa, Mexico   
3               BUS                        Texas, United States   
4               BUS                        Texas, United States   
..              ...                                         ...   
348             JUD                     Colorado, United States   
349             JUD                     Colorado, United States   
350             LEG                         Utah, United States   
351             MED                   California, United States   
352             MED  Monterey County, California, United States   

    ActionGeo_ADM1Code Actor1Geo_CountryCode  CAST(GoldsteinScale AS FLOAT)  \
0                 USNJ                    US                            7.0   
1                 USMS                    US                           -6.5   
2                 MX25                    US                            3.0   
3                 USTX                    US                            3.0   
4                 USTX                    US                            5.0   
..                 ...                   ...                            ...   
348               USCO                    US                           -2.0   
349               USCO                    US                           -5.0   
350               USUT                    US                           -5.0   
351               USCA                    US                           -5.0   
352               USCA                    US                           -5.0   

    action_state  
0             NJ  
1             MS  
2             25  
3             TX  
4             TX  
..           ...  
348           CO  
349           CO  
350           UT  
351           CA  
352           CA  

[314 rows x 11 columns]

This reduces number of events further to 314.

# Normalize Goldstein 

Goldstein scale represents impact of a particular event quantitatively. The column is rename to represent it properly.

In [115]:
df1 = df1.rename(columns={
    'CAST(GoldsteinScale AS FLOAT)': 'goldsteinscale',
    'CAST(SQLDATE AS INTEGER)': 'sqldate'
}
)

The Goldstein scale ranges between -10 to 10 in general. The range of variation for this specific dataset is obtained as follows:

In [116]:
print(df1.goldsteinscale.min())
print(df1.goldsteinscale.max())

-10.0
8.3


In [119]:
min_scale, max_scale = -10.000005, 10.000005
norm_gold = df2['goldsteinscale'].apply(lambda x: (x - min_scale) / (max_scale - min_scale))

df2['norm_scale'] = norm_gold
df2

GLOBALEVENTID   sqldate  MonthYear  Year Actor1Code Actor1Type1Code  \
0        929813256  20200518     202005  2020        COP             COP   
1        929813257  20200518     202005  2020        COP             COP   
2        929813265  20200610     202006  2020        BUS             BUS   
3        929813266  20200610     202006  2020        BUS             BUS   
4        929813267  20200610     202006  2020        BUS             BUS   
..             ...       ...        ...   ...        ...             ...   
348      929815208  20200617     202006  2020     USAJUD             JUD   
349      929815209  20200617     202006  2020     USAJUD             JUD   
350      929815210  20200617     202006  2020     USALEG             LEG   
351      929815222  20200617     202006  2020     USAMED             MED   
352      929815223  20200617     202006  2020     USAMED             MED   

                             ActionGeo_FullName ActionGeo_ADM1Code  \
0                     New Jersey, United States               USNJ   
1                    Mississippi, United States               USMS   
2                   Tamaulipas, Sinaloa, Mexico               MX25   
3                          Texas, United States               USTX   
4                          Texas, United States               USTX   
..                                          ...                ...   
348                     Colorado, United States               USCO   
349                     Colorado, United States               USCO   
350                         Utah, United States               USUT   
351                   California, United States               USCA   
352  Monterey County, California, United States               USCA   

    Actor1Geo_CountryCode  goldsteinscale action_state  norm_scale  
0                      US             7.0           NJ       0.850  
1                      US            -6.5           MS       0.175  
2                      US             3.0           25       0.650  
3                      US             3.0           TX       0.650  
4                      US             5.0           TX       0.750  
..                    ...             ...          ...         ...  
348                    US            -2.0           CO       0.400  
349                    US            -5.0           CO       0.250  
350                    US            -5.0           UT       0.250  
351                    US            -5.0           CA       0.250  
352                    US            -5.0           CA       0.250  

[314 rows x 12 columns]

# Aggregate data

To provide user with a useful metric, data are grouped by state, year, month, and actor type. For each group, the number of events are counted and normalized scales are summed to calculate an average scale.

In [122]:
df3 = df2.groupby(['action_state', 'Year', 'MonthYear',
                 'Actor1Type1Code'],
                as_index=False).agg({
                    "GLOBALEVENTID": ["count"],
                    "norm_scale": ["sum"]}).rename(columns={
                        'GLOBALEVENTID': 'events_count',
                        'norm_scale': 'norm_scale_sum'})
df3

action_state  Year MonthYear Actor1Type1Code events_count norm_scale_sum
                                                        count            sum
0                 2020    202006             BUS            2          1.190
1                 2020    202006             EDU            1          0.595
2                 2020    202006             GOV            1          0.800
3                 2020    202006             MED            1          0.400
4             01  2020    202006             BUS            1          0.670
..           ...   ...       ...             ...          ...            ...
104           VA  2020    202006             JUD            1          0.480
105           VA  2020    202006             LEG            6          3.290
106           WI  2020    202006             GOV            1          0.250
107           WI  2020    202006             JUD            2          0.900
108           WI  2020    202006             MED            1          0.650

[109 rows x 6 columns]

In [124]:
import re
pattern = re.compile("^[a-zA-Z]+$")
df4 = df3.loc[df.action_state.str.contains(pattern)]
df4

action_state  Year MonthYear Actor1Type1Code events_count norm_scale_sum
                                                        count            sum
0                 2020    202006             BUS            2          1.190
1                 2020    202006             EDU            1          0.595
3                 2020    202006             MED            1          0.400
4             01  2020    202006             BUS            1          0.670
7             AL  2020    202006             BUS            1          0.640
..           ...   ...       ...             ...          ...            ...
104           VA  2020    202006             JUD            1          0.480
105           VA  2020    202006             LEG            6          3.290
106           WI  2020    202006             GOV            1          0.250
107           WI  2020    202006             JUD            2          0.900
108           WI  2020    202006             MED            1          0.650

[96 rows x 6 columns]

In [125]:
df4.columns = df4.columns.droplevel(1)
df4

action_state  Year  MonthYear Actor1Type1Code  events_count  \
0                 2020     202006             BUS             2   
1                 2020     202006             EDU             1   
3                 2020     202006             MED             1   
4             01  2020     202006             BUS             1   
7             AL  2020     202006             BUS             1   
..           ...   ...        ...             ...           ...   
104           VA  2020     202006             JUD             1   
105           VA  2020     202006             LEG             6   
106           WI  2020     202006             GOV             1   
107           WI  2020     202006             JUD             2   
108           WI  2020     202006             MED             1   

     norm_scale_sum  
0             1.190  
1             0.595  
3             0.400  
4             0.670  
7             0.640  
..              ...  
104           0.480  
105           3.290  
106           0.250  
107           0.900  
108           0.650  

[96 rows x 6 columns]

# DataFrame to Dictionary

In [127]:
dict_df = df4.to_dict(orient='records')
print(dict_df)

[{'action_state': '', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'BUS', 'events_count': 2, 'norm_scale_sum': 1.1899999050000476}, {'action_state': '', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'EDU', 'events_count': 1, 'norm_scale_sum': 0.5949999525000238}, {'action_state': '', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'MED', 'events_count': 1, 'norm_scale_sum': 0.400000049999975}, {'action_state': '01', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'BUS', 'events_count': 1, 'norm_scale_sum': 0.6699999150000425}, {'action_state': 'AL', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'BUS', 'events_count': 1, 'norm_scale_sum': 0.639999930000035}, {'action_state': 'AZ', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'COP', 'events_count': 1, 'norm_scale_sum': 0.400000049999975}, {'action_state': 'AZ', 'Year': 2020, 'MonthYear': 202006, 'Actor1Type1Code': 'JUD', 'events_count': 1, 'norm_scale_sum': 0.2500001249999375}, {'action_state

# Load Database

In [131]:
from six.moves import configparser

config = configparser.ConfigParser()
# TODO: Make sure to read the correct config.ini file on AWS workers
config.read('C:\\Users\\arali\\Files\\Professional\\Self Study\\Data Sceince\\Portfolio\\GDELT Pipeline - Desktop\\jupyter\\config.ini')
dbname = config.get('dbauth', 'dbname')
dbuser = config.get('dbauth', 'user')
dbpass = config.get('dbauth', 'password')
dbhost = config.get('dbauth', 'host')
dbport = config.get('dbauth', 'port')

'5432'

In [ ]:
from sqlalchemy import create_engine

db = create_engine('postgres://%s%s/%s' % (dbuser, dbhost, dbname))
con = None
con = psycopg2.connect(
    database=dbname,
    host=dbhost,
    user=dbuser,
    password=dbpass)